src/libraries/System.Linq/src/System/Linq/Sequence.cs

```diff
@@ -0,0 +1,156 @@
+                if (typeof(T) == typeof(uint) && (range = TryUseRange<ulong>(start, endInclusive, step, uint.MaxValue)) is not null) return range;
+                if (typeof(T) == typeof(int) && (range = TryUseRange<long>(start, endInclusive, step, int.MaxValue)) is not null) return range;
+                if (typeof(T) == typeof(ulong) && (range = TryUseRange<UInt128>(start, endInclusive, step, ulong.MaxValue)) is not null) return range;
+                if (typeof(T) == typeof(long) && (range = TryUseRange<Int128>(start, endInclusive, step, long.MaxValue)) is not null) return range;
+                if (typeof(T) == typeof(nuint) && (range = TryUseRange<UInt128>(start, endInclusive, step, nuint.MaxValue)) is not null) return range;
+                if (typeof(T) == typeof(nint) && (range = TryUseRange<Int128>(start, endInclusive, step, nint.MaxValue)) is not null) return range;
+
+                // Otherwise, just produce an incrementing sequence.
+                return IncrementingIterator(start, endInclusive, step);
+            }
+            else if (step < T.Zero)
+            {
+                // step < 0. Validate that endInclusive <= start, as otherwise we can't easily guarantee that the sequence will terminate.
+                if (endInclusive > start)
+                {
+                    ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.endInclusive);
+                }
+
+                // Then produce the decrementing sequence.
+                return DecrementingIterator(start, endInclusive, step);
+            }
+            else
+            {
+                // step == 0. If start != endInclusive, then the sequence would be infinite. As such, we validate
+                // that they're equal, and if they are, we return a sequence that yields the start/endInclusive value once.
+                if (start != endInclusive)
+                {
+                    ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.step);
+                }
+
+                return Repeat(start, 1);
+            }
+
+            static RangeIterator<T>? TryUseRange<TLarger>(T start, T endInclusive, T step, TLarger maxValue) where TLarger : INumber<TLarger>
+            {
+                if (step == T.One &&
+                    TLarger.CreateTruncating(endInclusive) - TLarger.CreateTruncating(start) + TLarger.One <= maxValue)
+                {
+                    return new RangeIterator<T>(start, endInclusive + T.One);
+                }
+
+                return null;
+            }
+
+            static IEnumerable<T> IncrementingIterator(T start, T endInclusive, T step)
+            {
+                Debug.Assert(step > T.Zero);
+
+                yield return start;
+
+                while (true)
+                {
+                    T next = start + step;
+
+                    if (next >= endInclusive || next <= start)
```

Comment:
Should we be checking for saturation or reaching a value where `step` no longer produces a new result at all?

Reply:
That should be handled by the next <= start, right? (start was updated to be the last value)

Reply:
FWIW, I think confusion about `next <= start` has come up a few times.  Perhaps it would have been better if you'd introduced a `last` instead of just updating start in place :smile:

Reply:
Ah, yeah. I read that `start` was being updated, but my brain didn't connect that this meant it was effectively the last yielded value.

It would be clearer if we had something like `current` and `previous` (to avoid potential confusion of `last` with `end`).

A comment explicitly specifying we're handling `overflow` and `saturation` would also help, as it's the kind of thing that is easy to miss or forget given this can be any numeric type and people may only be familiar with typical integer semantics.